src/provider/mlkzg.rs

```diff
@@ -0,0 +1,621 @@
+        <NE::CE as CommitmentEngineTrait<NE>>::commit(ck, &polys[i])
+          .comm
+          .preprocessed()
+      })
+      .collect();
+
+    // Phase 2
+    // We do not need to add x to the transcript, because in our context x was
+    // obtained from the transcript.
+    let r = Self::compute_challenge(&C.comm.preprocessed(), eval, &com, transcript);
+    let u = vec![r, -r, r * r];
+
+    // Phase 3 -- create response
+    let mut com_all = com.clone();
+    com_all.insert(0, C.comm.preprocessed());
+    let (w, v) = kzg_open_batch(&com_all, &polys, &u, transcript);
+
+    Ok(EvaluationArgument { com, w, v })
+  }
+
+  /// A method to verify purported evaluations of a batch of polynomials
+  fn verify(
+    vk: &Self::VerifierKey,
+    transcript: &mut <NE as NovaEngine>::TE,
+    C: &Commitment<NE>,
+    point: &[E::Fr],
+    P_of_x: &E::Fr,
+    pi: &Self::EvaluationArgument,
+  ) -> Result<(), NovaError> {
+    let x = point.to_vec();
+    let y = P_of_x;
+
+    // vk is hashed in transcript already, so we do not add it here
+
+    let kzg_verify_batch = |vk: &KZGVerifierKey<E>,
+                            C: &Vec<E::G1Affine>,
+                            W: &Vec<E::G1Affine>,
+                            u: &Vec<E::Fr>,
+                            v: &Vec<Vec<E::Fr>>,
+                            transcript: &mut <NE as NovaEngine>::TE|
+     -> bool {
+      let k = C.len();
+      let t = u.len();
+
+      let q = Self::get_batch_challenge(C, u, v, transcript);
+      let q_powers = Self::batch_challenge_powers(q, k); // 1, q, q^2, ..., q^(k-1)
+
+      // Compute the commitment to the batched polynomial B(X)
+      let c_0: E::G1 = C[0].into();
+      let C_B = (c_0 + NE::GE::vartime_multiscalar_mul(&q_powers[1..k], &C[1..k])).preprocessed();
+
+      // Compute the batched openings
+      // compute B(u_i) = v[i][0] + q*v[i][1] + ... + q^(t-1) * v[i][t-1]
+      let B_u = (0..t)
+        .map(|i| {
+          assert_eq!(q_powers.len(), v[i].len());
+          q_powers
+            .iter()
+            .zip_eq(v[i].iter())
+            .map(|(a, b)| *a * *b)
+            .sum()
+        })
+        .collect::<Vec<E::Fr>>();
+
+      let d_0 = Self::verifier_second_challenge(&C_B, W, transcript);
+      let d = [d_0, d_0 * d_0];
+
+      assert!(t == 3);
+      // We write a special case for t=3, since this what is required for
+      // mlkzg. Following the paper directly, we must compute:
+      // let L0 = C_B - vk.G * B_u[0] + W[0] * u[0];
+      // let L1 = C_B - vk.G * B_u[1] + W[1] * u[1];
+      // let L2 = C_B - vk.G * B_u[2] + W[2] * u[2];
+      // let R0 = -W[0];
+      // let R1 = -W[1];
+      // let R2 = -W[2];
+      // let L = L0 + L1*d[0] + L2*d[1];
+      // let R = R0 + R1*d[0] + R2*d[1];
+      //
+      // We group terms to reduce the number of scalar mults (to seven):
+      // In Rust, we could use MSMs for these, and speed up verification.
+      let L = E::G1::from(C_B) * (E::Fr::ONE + d[0] + d[1])
```

Comment:
Instead of computing `C_B * (1 + d[0] + d[1])`, we should directly compute the MSM of `C` with `q_powers[..] * (1 + d[0] + d[1])`.

Reply:
Fixed in https://github.com/lurk-lab/arecibo/pull/231